src/main/java/org/apache/commons/lang3/StringUtils.java

```diff
@@ -7968,9 +7968,160 @@ public static String rotate(final String str, final int shift) {
             }
         }
         if (match || preserveAllTokens && lastMatch) {
-            list.add(str.substring(start, i));
+            buffer.add(str.substring(start, i));
+        }
+        return buffer.toArray();
+    }
+
+    private static final ThreadLocal<SplitBufferThreadLocalHelper> SPLIT_BUFFER_THREAD_LOCAL
+            = new ThreadLocal<SplitBufferThreadLocalHelper>() {
+        @Override
+        protected SplitBufferThreadLocalHelper initialValue() {
+            return new SplitBufferThreadLocalHelper();
+        }
+    };
+
+
+
+    /**
+     * Private class act as a buffer while splitting.
+     * "SplitBufferThreadLocalHelper" is constructed as a thread local variable so it is
```

Comment:
Yes, it is thread-safe, but as soon as project loom is released thread-safety will not be our concerns. It will not be fiber safe. Using thread-local variables instead of objects is a hack and will bite back as hard as static variables bit our assets when we started to create multi-thread programs.

Reply:
I think not only `ThreadLocal`, but also the behavior of any *mutable object* would be unpredictable if the execution of a method call could be break(i.e. another method takes over the thread during the method call).I guess there might be some "safe-point" where another method could have access to the execution in project loom?

Reply:
> Another relatively major design decision concerns thread locals. Currently, thread-local data is represented by the (Inheritable)ThreadLocal class(es). How do we treat thread-locals in fibers? Crucially, ThreadLocals have two very different uses. One is associating data with a thread context. Fibers will probably need this capability, too. Another is to reduce contention in concurrent data structures with striping. That use abuses ThreadLocal as an approximation of a processor-local (more precisely, a CPU-core-local) construct. With fibers, the two different uses would need to be clearly separated, as now a thread-local over possibly millions of threads (fibers) is not a good approximation of processor-local data at all. This requirement for a more explicit treatment of thread-as-context vs. thread-as-an-approximation-of-processor is not limited to the actual ThreadLocal class, but to any class that maps Thread instances to data for the purpose of striping. If fibers are represented by Threads, then some changes would need to be made to such striped data structures. In any event, it is expected that the addition of fibers would necessitate adding an explicit API for accessing processor identity, whether precisely or approximately.

See:[https://cr.openjdk.java.net/~rpressler/loom/Loom-Proposal.html](https://cr.openjdk.java.net/~rpressler/loom/Loom-Proposal.html)